rat/ops.py

```diff
@@ -108,15 +250,111 @@ def scalar(self):
     def get_key(self):
         return self.name
 
-    def code(self):
+    def fold(self):
+        if self.lower is not None:
+            self.lower = self.lower.fold()
+            if not isinstance(self.lower, (RealConstant, IntegerConstant)):
+                raise ConstantFoldError(
+                    f"Lower bound value must fold into a Numeric constant at compile time, but folded expression {self.lower} is not a constant!",
+                    self.line_index,
+                    self.column_index,
+                )
+
+        if self.upper is not None:
+            self.upper = self.upper.fold()
+            if not isinstance(self.upper, (RealConstant, IntegerConstant)):
+                raise ConstantFoldError(
+                    f"Upper bound value must fold into a Numeric constant at compile time, but folded expression {self.upper} is not a constant!",
+                    self.line_index,
+                    self.column_index,
+                )
+
+        if self.subscript:
+            self.subscript = self.subscript.fold()
+
+        return Param(
+            name=self.name,
+            prime=self.prime,
+            subscript=self.subscript,
+            lower=self.lower,
+            upper=self.upper,
+            assigned_by_scan=self.assigned_by_scan,
+            line_index=self.line_index,
+            column_index=self.column_index,
+        )
+
+    def __post_init__(self):
+        self.out_type = types.NumericType
+
+    def code(self, scalar=False):
         variable_code = self.variable.code()
         if self.subscript:
-            return f"parameters['{variable_code}'][{self.subscript.code()}]"
+            if self.assigned_by_scan:
+                if scalar == False:
+                    raise Exception("Internal Error: Cannot do a vector evaluation of an assigned variable")
+
+                return f"carry[index - {self.subscript.code(scalar)} - 1]"
```

Comment:
Assume that when this variable is marked that it will be assigned by a scan that there will be a variable `carry` available that will be saving the carry values. This assumes things are sorted in a certain way and whatnot.